Adding docs for test coverage (#7615)
microsoft/vscode-docs: docs/python/testing.md

3. Run your tests, which will now be run in parallel.

## Run tests with coverage

Test coverage is a measure of how much of your code is covered by your tests, which can help you identify areas of your code that are not being fully tested. For more information on test coverage, visit VS Code's [Test Coverage documentation](/docs/editor/testing#test-coverage.md).

To run tests with coverage, select the coverage run icon in the Test Explorer or “Run with coverage” from any menu you normally trigger test runs from. The Python extension will run coverage using the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin if you are using pytest, or with [`coverage.py`](http://coverage.py/) for unittest.
> **Note**: Before running tests with coverage, make sure to install the correct testing package for your project.

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.

Author: ```suggestion
Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line and branch coverage metrics for each file and folder in your workspace.

>**Note**: Test coverage may also appear as a "Test Coverage" sub-tab in the Test Explorer.
```

To run tests with coverage, select the coverage run icon in the Test Explorer or “Run with coverage” from any menu you normally trigger test runs from. The Python extension will run coverage using the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin if you are using pytest, or with [`coverage.py`](http://coverage.py/) for unittest.
> **Note**: Before running tests with coverage, make sure to install the correct testing package for your project.

Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.

> 
> ![]()

Author: ```suggestion
```

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a tab specifically for coverage will appear below the Testing tab, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line and branch coverage metrics for each file and folder in your workspace.

Author: ```suggestion
Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line and branch coverage metrics for each file and folder in your workspace.
```

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line and branch coverage metrics for each file and folder in your workspace.

Reviewer: need to remove references to `branch coverage metrics` as these are no longer a part of this release

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line and branch coverage metrics for each file and folder in your workspace.

Author: ```suggestion
Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`kbstyle(F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.
```

> For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` to include your specifications. When the pytest argument `—cov` exists in `python.testing.pytestArgs`, the Python extension will make no edits to coverage args to allow the user’s customizations to take effect. If there is no `-cov` argument found, the extension will add `-cov=.` and `-cov-branch` to the pytest args prior to run to enable branch level coverage at the workspace root.

Reviewer: remove " and `-cov-branch`", also change `prior to run to enable branch level coverage at the workspace root.` to `prior to run to enable coverage at the workspace root.`

> For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` to include your specifications. When the pytest argument `—cov` exists in `python.testing.pytestArgs`, the Python extension will make no edits to coverage args to allow the user’s customizations to take effect. If there is no `-cov` argument found, the extension will add `-cov=.` and `-cov-branch` to the pytest args prior to run to enable branch level coverage at the workspace root.

Author: ```suggestion
For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` to include your specifications. When the pytest argument `—cov` exists in `python.testing.pytestArgs`, the Python extension will make no edits to coverage args to allow the user’s customizations to take effect. If there is no `-cov` argument found, the extension will add `-cov=.` to the pytest args prior to run to enable coverage at the workspace root.
```

3. Run your tests, which will now be run in parallel.

## Run tests with coverage

> Test coverage is a measure of how much of your code is covered by your tests, which can help you identify areas of your code that are not being fully tested. For more information on test coverage, visit VS Code's [Test Coverage documentation](/docs/editor/testing#test-coverage.md).

Reviewer 2: ```suggestion
Test coverage is a measure of how much of your code is covered by your tests, which can help you identify areas of your code that are not being fully tested. For more information on test coverage, visit VS Code's [Test Coverage documentation](/docs/editor/testing#_test-coverage.md).
```

Reviewer 2: I'm not sure this is right but the link was broken to me. Going to the [URL ](https://code.visualstudio.com/docs/editor/testing#_test-coverage)it seemed it needed a `_` but it'd be good to double check

> To run tests with coverage, select the coverage run icon in the Test Explorer or “Run with coverage” from any menu you normally trigger test runs from. The Python extension will run coverage using the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin if you are using pytest, or with [`coverage.py`](http://coverage.py/) for unittest.

Reviewer 2: ```suggestion
To run tests with coverage enabled, select the coverage run icon in the Test Explorer or the “Run with coverage” option from any menu you normally trigger test runs from. The Python extension will run coverage using the [`pytest-cov`](https://pypi.org/project/pytest-cov/) plugin if you are using pytest, or with [`coverage.py`](https://coverage.readthedocs.io/) for unittest.
```

Reviewer 2: also not confident on the coverage.py link, just know that the previous one didn't work for me

> > **Note**: Before running tests with coverage, make sure to install the correct testing package for your project.

Reviewer 2: I wonder if this note could be misleading since you don't need to install unittest, as it's built-in?  Or maybe this could be tweaked to say "install the correct testing coverage package for your project"?

> > **Note**: Before running tests with coverage, make sure to install the correct testing package for your project.

Author: ```suggestion
> **Note**: Before running tests with coverage, make sure to install the correct testing coverage package for your project.
```

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. A Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.
> 
> >**Note**: Test coverage results may also appear as a "Test Coverage" sub-tab in the Test Explorer.

Author: ```suggestion
Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. Test coverage results will appear as a "Test Coverage" sub-tab in the Test Explorer, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.
```

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.

Reviewer 2: This is a nit suggestion so feel free to disregard it, but I think we could  simplify this paragraph by suggesting one way to view coverage (e.g. through the test explorer), as the other ones are covered in the VS Core core doc.

> Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. These can be closed and re-opened via the Run Results panel in the bottom where it says "Close Test Coverage" or "View Test Coverage" under the most recent test run. Additionally, a Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.

Author: ```suggestion
Once the coverage run is complete, lines will be highlighted in the editor for line level coverage. A Test Coverage tab will appear below the Testing tab in the Test Explorer, also with a beaker icon, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.
```

![Gif showing running Python tests with coverage.](/docs/python/images/testing/python-test-coverage.gif)

> For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` to include your specifications. When the pytest argument `—cov` exists in `python.testing.pytestArgs`, the Python extension will make no edits to coverage args to allow the user’s customizations to take effect. If there is no `-cov` argument found, the extension will add `-cov=.` to the pytest args prior to run to enable coverage at the workspace root.

Reviewer 2: ```suggestion
For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` to include your specifications. When the pytest argument `--cov` exists in `python.testing.pytestArgs`, the Python extension will make no edits to coverage args to allow the user’s customizations to take effect. If there is no `-cov` argument found, the extension will add `-cov=.` to the pytest args prior to run to enable coverage at the workspace root.
```
